src/main/java/com/aws/greengrass/mqtt/bridge/clients/LocalMqtt5Client.java

```diff
@@ -0,0 +1,391 @@
+            }
+
+        }
+    }
+    @Override
+    public boolean supportsTopicFilters() {
+        return true;
+    }
+
+    @Override
+    public void updateSubscriptions(Set<String> topics, Consumer<MqttMessage> messageHandler) {
+        this.messageHandler = messageHandler;
+
+        this.toSubscribeLocalMqttTopics = new HashSet<>(topics);
+        LOGGER.atDebug().kv("topics", topics).log("Updated local MQTT5 topics to subscribe");
+        if (client.getIsConnected()) {
+            updateSubscriptionsInternal(true);
+        }
+    }
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void updateSubscriptionsInternal(boolean sessionPresent) {
+        synchronized (subscribeLock) {
+            if (subscribeFuture != null) {
+                subscribeFuture.cancel(true);
+            }
+
+            if (!sessionPresent) {
+                // TODO: Work this out better
+                // Need to resubscribe to dropped topics
+                toSubscribeLocalMqttTopics.addAll(subscribedLocalMqttTopics);
+            }
+
+            Set<String> topicsToRemove = new HashSet<>(subscribedLocalMqttTopics);
+            topicsToRemove.removeAll(toSubscribeLocalMqttTopics);
+
+            topicsToRemove.forEach(s -> {
+                try {
+                    unsubscribe(s);
+                    LOGGER.atDebug().kv(TOPIC, s).log("Unsubscribed from topic");
+                    subscribedLocalMqttTopics.remove(s);
+                } catch (Exception e) {
+                    // TODO: Unsubscribe doesn't throw an exception
+                    LOGGER.atError().kv(TOPIC, s).setCause(e).log("Unable to unsubscribe");
+                    // If we are unable to unsubscribe, leave the topic in the set
+                    // so that we can try to remove next time.
+                }
+            });
+
+            Set<String> topicsToSubscribe = new HashSet<>(toSubscribeLocalMqttTopics);
+            topicsToSubscribe.removeAll(subscribedLocalMqttTopics);
+
+            LOGGER.atDebug().kv("topics", topicsToSubscribe).log("Subscribing to MQTT topics");
+
+            subscribeToTopics(topicsToSubscribe);
+
+        }
+    }
+
+    private void subscribe(String topic) {
+        if (client.getIsConnected()) {
+            SubscribePacket subscribePacket = new SubscribePacket.SubscribePacketBuilder()
+                    .withSubscription(topic, QOS).build();
+            LOGGER.atDebug().kv(TOPIC, topic).log("Subscribing to MQTT topic");
+            //TODO: deal with subscribeFuture
+            this.subscribeFuture = client.subscribe(subscribePacket);
+            try {
+                SubAckPacket subAckPacket = this.subscribeFuture.get();
+                //TODO: Fix logging
+                LOGGER.atDebug().kv("reason", subAckPacket.getReasonString()).log("while subscribing");
+            } catch (InterruptedException | ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)